File: Algorithm-CPP/44_마구간정하기.cpp
#include <iostream>
#include <algorithm>
using namespace std;

int main() {
    int n, c;
    cin >> n >> c;

    int arr[n];
    for(int i = 0; i < n; i++) cin >> arr[i];
    sort(arr, arr+n);

    int lt = 1, rt = arr[n-1], res = 0;
    while(lt <= rt) {
        int mid = (lt + rt) / 2;
        int pos = arr[0], cnt = 1;

        for(int i = 1; i < n; i++) {
            if(arr[i] - pos >= mid) {
                cnt++;
                pos = arr[i];
            }
        }

        if(cnt >= c) {
            lt = mid + 1;
            res = mid;
        } else rt = mid - 1;
    }

    cout << res << endl;

    return 0;
}

File: Algorithm-Swift/Algorithm-Swift/Inflearn/44_마구간정하기.swift
//
//  44_마구간정하기.swift
//  Algorithm-Swift
//
//  Created by SHIN YOON AH on 2023/08/20.
//

import Foundation

func solution44() {
    let nc = readLine()!.split(separator: " ").map { Int($0)! }
    let n = nc.first!, c = nc.last!
    var arr: [Int] = []
    (0..<n).forEach { _ in
        arr.append(Int(readLine()!)!)
    }
    let sortedArr = arr.sorted()

    var lt = 1, rt = sortedArr.last!, res = 0
    while lt <= rt {
        let mid = (lt + rt) / 2
        var pos = sortedArr[0], cnt = 1

        for i in 1..<n {
            if sortedArr[i] - pos >= mid {
                cnt += 1
                pos = sortedArr[i]
            }
        }

        if cnt >= c {
            lt = mid + 1
            res = mid
        } else {
            rt = mid - 1
        }
    }

    print(res)
}
